File: blaise-graphtheory3/src/main/java/com/googlecode/blaisemath/graph/mod/metrics/ClusteringCoefficient.java
package com.googlecode.blaisemath.graph.mod.metrics;

/*
 * #%L
 * BlaiseGraphTheory
 * --
 * Copyright (C) 2009 - 2018 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.google.common.graph.Graph;
import com.google.common.graph.Graphs;
import com.googlecode.blaisemath.graph.GraphUtils;

import java.util.Set;

/**
 * Global metric describing the clustering coefficient of the graph; in the
 * directed case, measures "transitivity", i.e. when a-%gt;b,b-%gt;c implies
 * a-%gt;c
 *
 * @author Elisha Peterson
 */
public class ClusteringCoefficient extends AbstractGraphMetric<Double> {

    public ClusteringCoefficient() {
        super("Clustering coefficient", "Computes the clustering coefficient:"
                + " Out of all triples of vertices with at least two edges, how many have three edges?", true);
    }

    @Override
    public Double apply(Graph graph) {
        int[] tri = triples(graph);
        int triangles = tri[0], triples = tri[1];
        if (!graph.isDirected()) {
            triangles /= 3;
            triples -= 2 * triangles;
        }
        return triangles / (double) triples;
    }

    /**
     * Computes triple characteristics of a graph.
     *
     * @param <V> coordinate type of graph
     * @param graph the graph
     * @return int[] array where first entry is number of triangles and second
     *      is number of path triples (i.e., when three nodes are connected together)
     */
    static <V> int[] triples(Graph<V> graph) {
        int triangles = 0;
        int triples = 0;
        for (V node : graph.nodes()) {
            Set<V> g1 = graph.adjacentNodes(node);
            int dist1 = g1.size();
            int aDist1 = Graphs.inducedSubgraph(graph, g1).edges().size();
            Set<V> g2 = GraphUtils.neighborhood(graph, node, 2);
            int dist2 = g2.size() - 1 - g1.size();

            if (graph.isDirected()) {
                // in the directed case, potential triples are connected nodes at distance 1 and nodes at distance 2
                // ... each node at distance 2 contributes a triple, but no triangle
                triples += aDist1 + dist2;
                triangles += aDist1;
            } else {
                // in copyUndirected case, each pair of nodes @ distance 1 contributes to a triple
                // ... each edge in this neighborhood indicates a triangle
                // corrections for later: each triangle is counted 3 times
                triples += dist1 * (dist1 - 1) / 2;
                triangles += aDist1;
            }
        }
        return new int[]{triangles, triples};
    }
}

File: blaise-graph-theory-kt/src/main/java/com/googlecode/blaisemath/graph/metrics/ClusteringCoefficient.kt
package com.googlecode.blaisemath.graph.metrics

import com.google.common.graph.Graph
import com.google.common.graph.Graphs
import com.googlecode.blaisemath.app.ApplicationMenuConfig
import com.googlecode.blaisemath.graph.GraphUtils
import com.googlecode.blaisemath.graph.app.AnimationUtils
import com.googlecode.blaisemath.graph.app.GraphApp
import com.googlecode.blaisemath.graph.app.GraphAppCanvas
import com.googlecode.blaisemath.graph.app.GraphAppFrameView
import com.googlecode.blaisemath.graph.app.MetricScaler
import com.googlecode.blaisemath.graph.generate.GraphGrowthRule
import com.googlecode.blaisemath.graph.generate.GraphSeedRule
import com.googlecode.blaisemath.graph.generate.HopGrowthRule
import com.googlecode.blaisemath.graph.layout.SpringLayoutPerformanceTest
import com.googlecode.blaisemath.graph.metrics.AdditiveSubsetMetricTest
import com.googlecode.blaisemath.graph.metrics.BetweenCentralityTest
import com.googlecode.blaisemath.graph.metrics.ClosenessCentralityTest
import com.googlecode.blaisemath.graph.metrics.CooperationMetric
import com.googlecode.blaisemath.graph.metrics.EigenCentralityTest
import com.googlecode.blaisemath.graph.metrics.GraphCentralityTest
import com.googlecode.blaisemath.graph.metrics.SubsetMetricsTest
import com.googlecode.blaisemath.graph.test.DynamicGraphTestFrame
import com.googlecode.blaisemath.graph.test.GraphTestFrame
import com.googlecode.blaisemath.graph.test.MyTestGraph
import com.googlecode.blaisemath.graphics.editor.AttributeSetPropertyModelTestFrame
import com.googlecode.blaisemath.graphics.svg.SvgElementGraphicConverter
import com.googlecode.blaisemath.graphics.svg.SvgGraphic
import com.googlecode.blaisemath.graphics.svg.SvgGraphicComponent
import com.googlecode.blaisemath.graphics.svg.SvgUtils
import com.googlecode.blaisemath.graphics.swing.AnchorTestFrame
import com.googlecode.blaisemath.graphics.testui.ContextMenuTestFrame
import com.googlecode.blaisemath.graphics.testui.HelloWorldTest
import com.googlecode.blaisemath.graphics.testui.PanAndZoomTestFrame
import com.googlecode.blaisemath.graphics.testui.SelectionTestFrame
import com.googlecode.blaisemath.graphics.testui.TooltipTestFrame
import com.googlecode.blaisemath.style.xml.AttributeSetAdapter
import com.googlecode.blaisemath.svg.HelloWorldSvg
import com.googlecode.blaisemath.svg.SvgCircle
import com.googlecode.blaisemath.svg.SvgCircle.CircleConverter
import com.googlecode.blaisemath.svg.SvgElement
import com.googlecode.blaisemath.svg.SvgElements
import com.googlecode.blaisemath.svg.SvgEllipse
import com.googlecode.blaisemath.svg.SvgEllipse.EllipseConverter
import com.googlecode.blaisemath.svg.SvgGroup
import com.googlecode.blaisemath.svg.SvgImage
import com.googlecode.blaisemath.svg.SvgImage.ImageConverter
import com.googlecode.blaisemath.svg.SvgIo
import com.googlecode.blaisemath.svg.SvgLine
import com.googlecode.blaisemath.svg.SvgLine.LineConverter
import com.googlecode.blaisemath.svg.SvgNamespaceFilter
import com.googlecode.blaisemath.svg.SvgPath
import com.googlecode.blaisemath.svg.SvgPath.SvgPathOperator
import com.googlecode.blaisemath.svg.SvgPathTest
import com.googlecode.blaisemath.svg.SvgPolygon
import com.googlecode.blaisemath.svg.SvgPolygon.PolygonConverter
import com.googlecode.blaisemath.svg.SvgPolyline
import com.googlecode.blaisemath.svg.SvgPolyline.PolylineConverter
import com.googlecode.blaisemath.svg.SvgRectangle
import com.googlecode.blaisemath.svg.SvgRectangle.RectangleConverter
import com.googlecode.blaisemath.svg.SvgRoot
import com.googlecode.blaisemath.svg.SvgRootTest
import com.googlecode.blaisemath.svg.SvgText
import com.googlecode.blaisemath.svg.SvgText.TextConverter
import com.googlecode.blaisemath.svg.SvgTool
import com.googlecode.blaisemath.test.AssertUtils
import com.googlecode.blaisemath.ui.PropertyActionPanel
import com.googlecode.blaisemath.util.Images
import junit.framework.TestCase
import org.junit.BeforeClass

/*
* #%L
* BlaiseGraphTheory
* --
* Copyright (C) 2009 - 2021 Elisha Peterson
* --
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* #L%
*/ /**
 * Global metric describing the clustering coefficient of the graph; in the
 * directed case, measures "transitivity", i.e. when a-%gt;b,b-%gt;c implies
 * a-%gt;c
 *
 * @author Elisha Peterson
 */
class ClusteringCoefficient : AbstractGraphMetric<Double?>("Clustering coefficient", "Computes the clustering coefficient:"
        + " Out of all triples of nodes with at least two edges, how many have three edges?", true) {
    override fun apply(graph: Graph?): Double? {
        val tri = triples<Any?>(graph)
        var triangles = tri.get(0)
        var triples = tri.get(1)
        if (!graph.isDirected()) {
            triangles /= 3
            triples -= 2 * triangles
        }
        return triangles / triples as Double
    }

    companion object {
        /**
         * Computes triple characteristics of a graph.
         *
         * @param <N> coordinate type of graph
         * @param graph the graph
         * @return int[] array where first entry is number of triangles and second
         * is number of path triples (i.e., when three nodes are connected together)
        </N> */
        fun <N> triples(graph: Graph<N?>?): IntArray? {
            var triangles = 0
            var triples = 0
            for (node in graph.nodes()) {
                val g1 = graph.adjacentNodes(node)
                val dist1 = g1.size
                val aDist1 = Graphs.inducedSubgraph(graph, g1).edges().size
                val g2 = GraphUtils.neighborhood(graph, node, 2)
                val dist2 = g2.size - 1 - g1.size
                if (graph.isDirected()) {
                    // in the directed case, potential triples are connected nodes at distance 1 and nodes at distance 2
                    // ... each node at distance 2 contributes a triple, but no triangle
                    triples += aDist1 + dist2
                    triangles += aDist1
                } else {
                    // in copyUndirected case, each pair of nodes @ distance 1 contributes to a triple
                    // ... each edge in this neighborhood indicates a triangle
                    // corrections for later: each triangle is counted 3 times
                    triples += dist1 * (dist1 - 1) / 2
                    triangles += aDist1
                }
            }
            return intArrayOf(triangles, triples)
        }
    }
}